### PR TITLE
Added GatherElements to Nuphar

### PR DESCRIPTION
**Pull request by yangchen-MS**
This change added GatherElements (op_ver 11) to the Nuphar provider.

### CODE REVIEW

File: onnxruntime/core/codegen/mti/tensor/gather_elements.cc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                           int64_t axis,
+                           const tvm::Tensor& indices,
+                           const std::string& name) {
+  tvm::Array<tvm::Expr> output_shape;
+  int64_t indices_rank = static_cast<int64_t>(indices->shape.size());
+  // output shape is the same as indices
+  for (int64_t i = 0; i < indices_rank; ++i)
+    output_shape.push_back(indices->shape[i]);
+
+  tvm::Expr idx_upper_bound = t->shape[axis];
```
Comment (KeDengMS) [inline onnxruntime/core/codegen/mti/tensor/gather_elements.cc]: >idx_upper_bound [](start = 12, length = 15)

make this dim -1, so we can save some compute in the loop below. And the upper_bound would be inclusive

File: onnxruntime/core/codegen/mti/tensor/gather_elements.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+#include "core/codegen/mti/tensor/gather_elements.h"
+
+#include "core/codegen/mti/mti_tvm_utils.h"
+#include <topi/transform.h>
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+tvm::Tensor GatherElements(const tvm::Tensor& t,
+                           int64_t axis,
```
Comment (KeDengMS) [inline onnxruntime/core/codegen/mti/tensor/gather_elements.cc]: >axis [](start = 35, length = 4)

can this be negative? if not, maybe should use unsigned types?

File: onnxruntime/test/providers/cpu/tensor/gather_elements_op_test.cc [hunk 3]
```diff
@@ -158,7 +161,10 @@ void RunTypedTest<std::string>() {
   test4.AddOutput<std::string>("output", {2, 2},
                                {"a", "a",
                                 "d", "d"});
-  test4.Run(OpTester::ExpectResult::kExpectFailure, "GatherElements op: Value in indices must be within bounds [-2 , 1]. Actual value is -3");
+  // skip nuphar, which will not throw error message but will ensure no out-of-bound access
```
Comment (KeDengMS) [inline onnxruntime/test/providers/cpu/tensor/gather_elements_op_test.cc]: >which will not throw error message but will ensure no out-of-bound access [](start = 18, length = 73)

I believe this should be the default behavior in ONNX spec for this op. Lots of devices won't be able to throw on invalid indices. Please open an issue to ONNX.

File: onnxruntime/core/codegen/mti/tensor/gather_elements.cc [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        tvm::Array<tvm::Expr> idx_vars;
+        for (int64_t j = 0; j < indices_rank; j++)
+          idx_vars.push_back(ovars[j]);
+        tvm::Expr idx = indices(idx_vars);
+        // when idx >= 0, we take tvm::max(..., 0), because (idx < 0) is 0
+        // when idx < 0, we take idx_upper_bound + tvm::max(...), because tvm::max(idx, 0) is 0
+        tvm::Expr real_idx = tvm::max(tvm::min(idx, idx_upper_bound - 1), 0) +
+                             (idx < 0) * (idx_upper_bound + tvm::max(idx, -idx_upper_bound));
+        // tvm idx must be of Int(32)
+        ivars.push_back(tvm::cast(tvm::Int(32), real_idx));
```
Comment (KeDengMS) [inline onnxruntime/core/codegen/mti/tensor/gather_elements.cc]: maybe better to make it a common util for index clamping, as other ops may reuse it too (Gather?)

File: onnxruntime/core/codegen/mti/tensor/gather_elements.cc (hunk 2 above)
Comment (yangchen-MS) [reply onnxruntime/core/codegen/mti/tensor/gather_elements.cc]: We always use int64_t for axis in other tensor apis. Seems it's better to follow the convention.

File: onnxruntime/test/providers/cpu/tensor/gather_elements_op_test.cc (hunk 3 above)
Comment (yangchen-MS) [reply onnxruntime/test/providers/cpu/tensor/gather_elements_op_test.cc]: OK

File: onnxruntime/core/codegen/mti/tensor/gather_elements.cc (hunk 4 above)
Comment (yangchen-MS) [reply onnxruntime/core/codegen/mti/tensor/gather_elements.cc]: Done.

File: onnxruntime/core/codegen/mti/mti_tvm_utils.h [hunk 5]
```diff
@@ -60,6 +60,9 @@ inline int64_t HandleNegativeAxis(int64_t axis, int64_t rank) {
   return axis = axis < 0 ? (axis + rank) : axis;
 }
 
+// Make sure idx is within the range of [-bound, bound - 1]
+tvm::Expr SafeIndex(const tvm::Expr& idx, const tvm::Expr& bound);
```
Comment (KeDengMS) [inline onnxruntime/core/codegen/mti/mti_tvm_utils.h]: >SafeIndex [](start = 10, length = 9)

Clamp index to be more exact

File: onnxruntime/core/codegen/mti/mti_tvm_utils.cc [hunk 6]
```diff
@@ -192,5 +192,13 @@ tvm::Array<tvm::Tensor> MakeInputsForExtern(const tvm::Array<tvm::Tensor>& input
   return fixed_inputs;
 }
 
+// Make sure idx is within the range of [-bound, bound - 1]
```
Comment (KeDengMS) [inline onnxruntime/core/codegen/mti/mti_tvm_utils.cc]: >within [](start = 20, length = 6)

clamped in

File: onnxruntime/core/codegen/mti/mti_tvm_utils.h (hunk 5 above)
Comment (yangchen-MS) [reply onnxruntime/core/codegen/mti/mti_tvm_utils.h]: Fixed. Thanks

File: onnxruntime/core/codegen/mti/mti_tvm_utils.cc (hunk 6 above)
Comment (yangchen-MS) [reply onnxruntime/core/codegen/mti/mti_tvm_utils.cc]: Done.

**Pull request merged**